### ISSUE TITLE
infer object from given array

### ISSUE THREAD

**Issue by nunomaduro** [vimeo/psalm#2398]
With psalm is possible to infer the object type from the given array?

```php
    /**
     * @template TKey as array-key
     * @template TValue
     *
     * @param  array<TKey, TValue>  $signature
     *
     * @return object{TKey: TValue}
     */
    function fromArray(array $signature)
    {
        return (object) $signature;
    }
```

**Comment by weirdan** [comment]
https://psalm.dev/r/acb3df9743 — sees to be working as you would expect it to.

**Comment by muglug** [comment]
Yeah, definitely broken

**Comment by weirdan** [comment]
@muglug note that the linked snippet does not illustrate the original request, as it does not call the function (my bad).

**Comment by muglug** [comment]
In https://psalm.dev/r/9344aeebab the template params are leaking

**Comment by nunomaduro** [comment]
Better example on what do I need: https://psalm.dev/r/b516634caf. Normally the param `foo` should be inferred to a string.

**Comment by muglug** [comment]
Specifically the TValue param is leaking. Psalm doesn’t support templated conversion of object-like arrays to stdClass, but it should support simple direct `(object)` converting object-like arrays to objects-with-properties

**Comment by muglug** [comment]
I fixed the leaking param in ccd6718

**Comment by muglug** [comment]
But I'm closing this, as creating templated `stdClass` instances is not something that I think is particularly useful, and it would require new docblock syntax.
